Fix npm ls for filters on prerelease packages

zkat opened the pull request:
treat a plain package name in ls as a versionless match. This fixes https://github.com/npm/npm/issues/9436

zkat committed:
test: adding some basic npm ls tests

Credit: @zkat
PR-URL: https://github.com/npm/npm/pull/12685

zkat committed:
ls: fix filter when prerelease version present

If `ls` for a package without using a semver filter,
for example, `npm ls foo` vs `foo@1.2.3`, `ls` was using semver
ranges based on `*`, which doesn't match prerelease versions.

So, if you had installed a prerelease version (`foo@1.2.3-xyz`),
the `npm ls` will return no results for `foo`, at all.

This patch bypasses the semver check entirely when there's no semver
filter for the search.

Fixes: https://github.com/npm/npm/issues/9436
Credit: @zkat
PR-URL: https://github.com/npm/npm/pull/12685

othiym23 commented:
Love the tests, and the change looks sound. Could you add a few comments explaining what's going on in that chunk of the code so that in the future we know which pieces are being used for that ls, and why? I read #9436 and the tests, and it still took me a little while to understand the purpose of this patch. (Also, a little context in the commit message would be helpful.) Aside from that, LGTM.

zkat commented:
(for record-keeping, @othiym23 gave this a 🐑 on slack)

zkat committed:
fstream@1.0.9

Partially addresses a race condition that caused missing
files during publish.

Credit: @evanlucas
PR-URL: https://github.com/npm/fstream/pull/50

zkat committed:
fstream-npm@1.1.0

`fstream-npm` always includes NOTICE files now.

Credit: @kemitchell
PR-URL: https://github.com/npm/fstream-npm/pull/17

zkat committed:
hosted-git-info@2.1.5

zkat committed:
inflight@1.0.5

zkat committed:
init-package-json@1.9.4

Stop using `package` for a variable, which defeats some bundlers and linters.

Credit: @adius
PR-URL: https://github.com/npm/init-package-json/pull/62

zkat committed:
readable-stream@2.1.3

Node 6 build and buffer updates
Credit: @calvinmetcalf

zkat committed:
which@1.2.9

Fix for paths startin with ../

Credit: @isaacs

zkat commented:
@evanlucas I just didn't rebase and repush this branch properly -- `fstream@1.0.9` isn't actually included here. That's just cruft from when I was merging it (oops, sorry!)

The `npm ls`-specific patches for this were included in `npm@3.9.3`
